src/xdp/programinspect.c

```diff
@@ -1065,3 +1069,204 @@ XdpInspect(
 
     return Action;
 }
+
+//
+// Control path routines.
+//
+
+VOID
+XdpProgramDeleteRule(
+    _Inout_ XDP_RULE *Rule
+    )
+{
+    if (Rule->Match == XDP_MATCH_IPV4_UDP_PORT_SET ||
+        Rule->Match == XDP_MATCH_IPV6_UDP_PORT_SET ||
+        Rule->Match == XDP_MATCH_IPV4_TCP_PORT_SET ||
+        Rule->Match == XDP_MATCH_IPV6_TCP_PORT_SET) {
+        XdpProgramReleasePortSet(&Rule->Pattern.IpPortSet.PortSet);
+    }
+
+    if (Rule->Match == XDP_MATCH_UDP_PORT_SET) {
+        XdpProgramReleasePortSet(&Rule->Pattern.PortSet);
+    }
+
+    if (Rule->Action == XDP_PROGRAM_ACTION_REDIRECT) {
+
+        switch (Rule->Redirect.TargetType) {
+
+        case XDP_REDIRECT_TARGET_TYPE_XSK:
+            if (Rule->Redirect.Target != NULL) {
+                XskDereferenceDatapathHandle(Rule->Redirect.Target);
+                Rule->Redirect.Target = NULL;
+            }
+            break;
+
+        default:
+            ASSERT(FALSE);
+        }
+    }
+}
+
+_Success_(TRUE)
```

Comment:
I'm pretty sure the correct way to do this is:
```
_Success_(return == TRUE)
```
or 
```
_Success_(return != FALSE)
```

Reply:
This is as intended: the function succeeds (i.e. initializes the `_Out_` parameters) independent of the `return` value.

Reply:
Oh, I see. Thanks for clarifying.

Reply:
then it's equivalent to not using the `_Success_` annotation at all.

Reply:
Is it?

```
typedef _Return_type_success_(return >= 0) LONG NTSTATUS;
```

The default for a function that returns an `NTSTATUS` appears to be different than this.

Reply:
That said, I agree it's a bit of a code smell and I've removed this annotation.